### ISSUE TITLE
False positives when used on a dating app

### ISSUE THREAD

**Issue by duogenesis** [bumble-tech/private-detector#17]
Thanks for the model! I'm using it on a small dating app which currently has 365,359 user-uploaded images. For my use case, I've noticed a lot of high scores for SFW images. For example, I've attached 5 images from the top 10 most NSFW images on @duolicious as predicted by `private-detector`. The model says that these have about a 79% chance of being NSFW despite them being fairly benign. I haven't cherry-picked these images; None of the other 5 images contain nudity either, though one of the other images is suggestive.

Although I haven't done any formal benchmarks, the accuracy of the model seems good for 50-50 splits of NSFW/SFW images. Though on data sets where the images are predominantly SFW, the proportion of false positives is quite high. I'd be interested to know how Bumble deals with this problem. Do you ensemble `private-detector` with other models? Do you use the model as a first-pass and leave the final decision up to a human moderator?

Interestingly, I've found that `private-detector` tends to give high NSFW probabilities to animals, especially cats. That makes me wonder how the training data for this model was obtained. Presumably a significant portion of training images come from in-app reports of users who sent cat pictures to make double entendres. I wonder if the accuracy of the model could be improved with cleaner training data.

**Edit**: It's possible that I haven't encoded the data properly before passing it to the model. Although, like I said, I've found that the model gives good results for 50-50 splits of lewd and inoffensive images, so if the implementation I'm using is wrong, it can't be _that_ wrong. The predictions are also highly invariant to manipulations of the input images, especially rotation. If you'd like to see my implementation to reproduce my results, it can be found [here](https://github.com/duolicious/duolicious-backend/blob/bc8154b64537f3d226108c5cac6cc3108436d8ae/antiporn/__init__.py#L100). Although it's part of a bigger application, everything in the `antiporn/` directory can be run separately. The `predict_nsfw` function is the entry point. It takes the raw image data (as read from disk) as a list of `BytesIO` objects.

---

![450-1d7b7114d1c0ff3de40f34738ea6e08bd7ca877942386eac7c444a2791d21f6f](https://github.com/user-attachments/assets/5b8c2a16-f94e-42b9-a3b0-f106db8a3966)
![450-6cda4a49ffe28e4b956330c9ff2808788d17274dbbf42847e8c154b83274c7ba](https://github.com/user-attachments/assets/32833d7c-407b-4d0d-b310-048edfdfa39b)
![450-ba0fcd92ed25ddf71f0054f4164e01312ae70095ffe0dc4b30e72ceefac12234](https://github.com/user-attachments/assets/4e5e77ce-0176-4d7a-9ef3-4138a0f50ef3)
![450-04b7e40216ae5397913f9561e7e41a8de7e548f2543856682f5cca19dd1081e2](https://github.com/user-attachments/assets/084a5c97-16f3-4565-9911-5a15ae392854)
![450-a2a9f12065cc30e13062cd042adc0a6282f5bcdf239b0e2389fa658248e74ccd](https://github.com/user-attachments/assets/5ef2962d-f8ea-41c2-b76a-40e132edfeb5)

**Comment by Steeeephen** [comment]
Hey! Sorry about the delay in responding - saw this over the christmas break but forgot to reply

> Though on data sets where the images are predominantly SFW, the proportion of false positives is quite high. I'd be interested to know how Bumble deals with this problem. Do you ensemble private-detector with other models? Do you use the model as a first-pass and leave the final decision up to a human moderator?

A couple of things here

* The open source PD is not the exact model that we use behind the scenes - we wanted to avoid adversarial testing so we trained a new model, very similar but some slight differences to make sure that we can still keep our users safe
* The inhouse version supports the private detector specifically, which is fully user-facing and separate from our moderation systems. Instead, we blur the photo sent and the receiver gets a warning and has the option to unblur the photo - so a false positive here isn't the end of the world

> That makes me wonder how the training data for this model was obtained. Presumably a significant portion of training images come from in-app reports of users who sent cat pictures to make double entendres. I wonder if the accuracy of the model could be improved with cleaner training data.

I can't go too in-depth in the makeup of the dataset but we do have human moderators in the loop who help us build our training sets, and we tend to lean more on human labels when building models for the safety team (where model performance is far more important). For the animal photos, I'm guessing that the model just hasn't seen many examples of those types of photos and doesn't really know how to deal with them - most photos on dating apps are (thankfully) just photos of people doing things, and we tend to [moderate a decent chunk of photos outside of that scope](https://bumble.com/en/guidelines). Generally setting a high threshold covers most of these edge cases though

Could definitely improve performance with cleaner training data, there are a few companies who reached out to us about their experience using the model and many said that finetuning the model on a dataset more aligned with their use case makes a much better setup for them

**Comment by duogenesis** [comment]
Awesome! I might give fine-tuning a shot. Thanks again for open-sourcing the model and for your response.
